File: backtracking/sol/subsets-II.cpp
class Solution {
public:
    vector<vector<int>> result;
    void fun(vector<int>& nums , int i, vector<int> list){
        if(i == nums.size()){
            result.push_back(list);
            return;
        }
        // let nums[i] = a;
        // take 'a'.
        list.push_back(nums[i]);
        fun(nums , i + 1 , list);
        list.pop_back();
        // not take 'a' . . . skip duplicates.
        i++;
        while(i < nums.size() && nums[i-1] == nums[i]){
            i++;
        }
        fun(nums , i , list);
        return;
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin() , nums.end());
        vector<int> vec;
        fun(nums , 0 , vec);
        return result;
    }
};

File: backtracking/sol/subsets-II.java
// TC -> 2^N (recursion)
// SC -> N (recursive stack space)
import java.util.*;

class Solution {
    List<List<Integer>> result = new ArrayList<>();
    void fun(int[] nums, int i, List<Integer> list) {
        if (i == nums.length) {
            result.add(new ArrayList<>(list));
            return;
        }
        // let nums[i] = a;
        // take 'a'.
        list.add(nums[i]);
        fun(nums, i + 1, list);
        list.remove(list.size() - 1);

        // not take 'a' . . . skip duplicates.
        i++;
        while (i < nums.length && nums[i - 1] == nums[i]) {
            i++;
        }
        fun(nums, i, list);
        return;
    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<Integer> vec = new ArrayList<>();
        fun(nums, 0, vec);
        return result;
    }
}
